<H1>Deep Learning Assignement

In [3]:
import matplotlib.pyplot as plt
import os
import re
import shutil
import string
import tensorflow as tf
import pandas as pd
import numpy as np

from tensorflow.keras import layers
from tensorflow.keras import losses
from tensorflow.keras.layers.experimental.preprocessing import TextVectorization
from sklearn.model_selection import train_test_split
from nltk.stem import WordNetLemmatizer
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from tensorflow.keras import layers, regularizers
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Federico\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [4]:
tf.random.set_seed(16)

In [5]:
import pickle


file_name = "test_splits_clean.pkl"

# Open the file for reading
with open(file_name, 'rb') as file_obj:
    X_test_clean, y_test_clean = pickle.load(file_obj)

In [6]:
file_name = "test_splits_unclean.pkl"

# Open the file for reading
with open(file_name, 'rb') as file_obj:
    X_test_unclean, y_test_unclean = pickle.load(file_obj)

Test CNN models

In [14]:
#load CNN models
CNN_model = tf.keras.models.load_model('CNN_model')
CNN_Tanh = tf.keras.models.load_model('CNN_model_Tanh')
CNN_Sigmoid = tf.keras.models.load_model('CNN_model_Sigmoid')


CLEAN DATA

In [15]:
#eval the models
loss, accuracy = CNN_model.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_Tanh.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Tanh accuracy: {accuracy} , loss: {loss}')

loss, accuracy = CNN_Sigmoid.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Sigmoid accuracy: {accuracy}, loss: {loss}')


5729/5729 [==============================] - 22s 4ms/step - loss: 0.4529 - accuracy: 0.8706
CNN model accuracy: 0.8705561757087708, loss: 0.4528604745864868
5729/5729 [==============================] - 21s 4ms/step - loss: 0.2642 - accuracy: 0.8980
CNN model with Tanh accuracy: 0.8980443477630615 , loss: 0.2642250657081604
5729/5729 [==============================] - 27s 5ms/step - loss: 0.2542 - accuracy: 0.8995
CNN model with Sigmoid accuracy: 0.8994626998901367, loss: 0.2541605532169342


| Model                      | Accuracy | Loss      |
|----------------------------|----------|-----------|
| CNN model with Relu        | 0.8706   | 0.4529    |
| CNN model with Tanh        | 0.8980   | 0.2642    |
| CNN model with Sigmoid     | 0.8995   | 0.2542    |


UNCLEAN DATA

In [18]:
#eval models on unclean data
loss, accuracy = CNN_model.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_Tanh.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Tanh accuracy: {accuracy} , loss: {loss}')

loss, accuracy = CNN_Sigmoid.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Sigmoid accuracy: {accuracy}, loss: {loss}')

5729/5729 [==============================] - 22s 4ms/step - loss: 6.3699 - accuracy: 0.3630
CNN model with Relu accuracy: 0.36304721236228943, loss: 6.369899272918701
5729/5729 [==============================] - 23s 4ms/step - loss: 4.1985 - accuracy: 0.3709
CNN model with Tanh accuracy: 0.37091344594955444 , loss: 4.1985249519348145
5729/5729 [==============================] - 24s 4ms/step - loss: 3.9936 - accuracy: 0.3626
CNN model with Sigmoid accuracy: 0.36255624890327454, loss: 3.9935927391052246


| Model                      | Accuracy | Loss     |
|----------------------------|----------|----------|
| CNN model with Relu        | 0.3630   | 6.3699   |
| CNN model with Tanh        | 0.3709   | 4.1985   |
| CNN model with Sigmoid     | 0.3626   | 3.9936   |


Testing LSTM models

In [12]:
#load the  LSTM models
LSTM_model = tf.keras.models.load_model('LSTM_model')
LSTM_Bi_model = tf.keras.models.load_model('LSTM_Bi_model')
LSTM_Attention_model = tf.keras.models.load_model('LSTM_Attention_model')
LSTM_Attention_Between_model = tf.keras.models.load_model('LSTM_Attention_Between_model')
LSTM_Attention_Best_model = tf.keras.models.load_model('LSTM_Attention_Best_model')


CLEAN DATA

In [16]:
#testing the models

# Evaluate the LSTM model
loss, accuracy = LSTM_model.evaluate(X_test_clean, y_test_clean)
print(f'LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the Bidirectional LSTM model
loss, accuracy = LSTM_Bi_model.evaluate(X_test_clean, y_test_clean)
print(f'Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_model.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Between_model.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention in Between model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Best_model.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention Best hyperparameters model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 167s 29ms/step - loss: 0.4200 - accuracy: 0.8735
LSTM model - Loss: 0.420012891292572, Accuracy: 0.8734582662582397
5729/5729 [==============================] - 209s 36ms/step - loss: 0.2115 - accuracy: 0.9105
Bidirectional LSTM model - Loss: 0.21147337555885315, Accuracy: 0.9105474352836609
5729/5729 [==============================] - 288s 50ms/step - loss: 0.3351 - accuracy: 0.8827
LSTM with Attention model - Loss: 0.33508560061454773, Accuracy: 0.882699191570282
5729/5729 [==============================] - 313s 55ms/step - loss: 0.3414 - accuracy: 0.8806
LSTM with Attention in Between model - Loss: 0.3413842022418976, Accuracy: 0.8806480765342712
5729/5729 [==============================] - 336s 59ms/step - loss: 0.3455 - accuracy: 0.8802
LSTM with Attention Best hyperparameters model - Loss: 0.3455374240875244, Accuracy: 0.8802225589752197


| Model                                  | Accuracy | Loss     |
|----------------------------------------|----------|----------|
| LSTM model                             | 0.8735   | 0.4200   |
| Bidirectional LSTM model               | 0.9105   | 0.2115   |
| LSTM with Attention model               | 0.8827   | 0.3351   |
| LSTM with Attention in Between model   | 0.8806   | 0.3414   |
| LSTM with Attention Best hyperparameters model | 0.8802 | 0.3455   |


UNCLEAN DATA

In [19]:
# Evaluate the LSTM model on unclean data
loss, accuracy = LSTM_model.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the Bidirectional LSTM model on unclean data
loss, accuracy = LSTM_Bi_model.evaluate(X_test_unclean, y_test_unclean)
print(f'Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model on unclean data
loss, accuracy = LSTM_Attention_model.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model on unclean data
loss, accuracy = LSTM_Attention_Between_model.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention in Between model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model on unclean data
loss, accuracy = LSTM_Attention_Best_model.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention Best hyperparameters model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 138s 24ms/step - loss: 4.2375 - accuracy: 0.3629
LSTM model - Loss: 4.2374958992004395, Accuracy: 0.3629271984100342
5729/5729 [==============================] - 182s 32ms/step - loss: 3.8392 - accuracy: 0.3651
Bidirectional LSTM model - Loss: 3.8392131328582764, Accuracy: 0.36514195799827576
5729/5729 [==============================] - 228s 40ms/step - loss: 3.7090 - accuracy: 0.3668
LSTM with Attention model - Loss: 3.7089524269104004, Accuracy: 0.3667948544025421
5729/5729 [==============================] - 420s 73ms/step - loss: 3.7730 - accuracy: 0.3646
LSTM with Attention in Between model - Loss: 3.773002862930298, Accuracy: 0.3646128177642822
5729/5729 [==============================] - 573s 100ms/step - loss: 3.7341 - accuracy: 0.3658
LSTM with Attention Best hyperparameters model - Loss: 3.734067440032959, Accuracy: 0.36580750346183777


| Model                                  | Accuracy | Loss     |
|----------------------------------------|----------|----------|
| LSTM model                             | 0.3629   | 4.2375   |
| Bidirectional LSTM model               | 0.3651   | 3.8392   |
| LSTM with Attention model              | 0.3668   | 3.7090   |
| LSTM with Attention in Between model   | 0.3646   | 3.7730   |
| LSTM with Attention Best hyperparameters model | 0.3658 | 3.7341   |


Lets evaluate the models when trained with unclean data

CNN models

In [20]:
#load CNN models
CNN_model_unclean = tf.keras.models.load_model('CNN_model2')
CNN_Tanh_unclean = tf.keras.models.load_model('CNN_model_Tanh2')
CNN_Sigmoid_unclean = tf.keras.models.load_model('CNN_model_Sigmoid2')

In [21]:
#eval the models
loss, accuracy = CNN_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_Tanh_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Tanh accuracy: {accuracy} , loss: {loss}')

loss, accuracy = CNN_Sigmoid_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN model with Sigmoid accuracy: {accuracy}, loss: {loss}')

5729/5729 [==============================] - 24s 4ms/step - loss: 1.0456 - accuracy: 0.7812
CNN model with Relu accuracy: 0.781239926815033, loss: 1.0455505847930908
5729/5729 [==============================] - 23s 4ms/step - loss: 0.2461 - accuracy: 0.9025
CNN model with Tanh accuracy: 0.9025120735168457 , loss: 0.24609433114528656
5729/5729 [==============================] - 23s 4ms/step - loss: 0.2584 - accuracy: 0.8991
CNN model with Sigmoid accuracy: 0.8990644812583923, loss: 0.2583971917629242


| Model                      | Accuracy | Loss     |
|----------------------------|----------|----------|
| CNN model with Relu        | 0.7812   | 1.0456   |
| CNN model with Tanh        | 0.9025   | 0.2461   |
| CNN model with Sigmoid     | 0.8991   | 0.2584   |


In [22]:
#evasl these woith clean data
loss, accuracy = CNN_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_Tanh_unclean.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Tanh accuracy: {accuracy} , loss: {loss}')

loss, accuracy = CNN_Sigmoid_unclean.evaluate(X_test_clean, y_test_clean)
print(f'CNN model with Sigmoid accuracy: {accuracy}, loss: {loss}')

5729/5729 [==============================] - 23s 4ms/step - loss: 3.5970 - accuracy: 0.3276
CNN model with Relu accuracy: 0.32760003209114075, loss: 3.5969834327697754
5729/5729 [==============================] - 24s 4ms/step - loss: 2.8005 - accuracy: 0.3353
CNN model with Tanh accuracy: 0.3352535367012024 , loss: 2.8005011081695557
5729/5729 [==============================] - 24s 4ms/step - loss: 2.9789 - accuracy: 0.3258
CNN model with Sigmoid accuracy: 0.3257780373096466, loss: 2.97892427444458


| Model                      | Accuracy | Loss     |
|----------------------------|----------|----------|
| CNN model with Relu        | 0.3276   | 3.5970   |
| CNN model with Tanh        | 0.3353   | 2.8005   |
| CNN model with Sigmoid     | 0.3258   | 2.9789   |


LSTM models

In [26]:
#load the  LSTM models
LSTM_model_unclean = tf.keras.models.load_model('LSTM_model2')
LSTM_Bi_model_unclean = tf.keras.models.load_model('LSTM_Bi_model2')
LSTM_Attention_model_unclean = tf.keras.models.load_model('LSTM_Attention_model2')
LSTM_Attention_Between_model_unclean = tf.keras.models.load_model('LSTM_Attention_Between_model2')
LSTM_Attention_Best_model_unclean = tf.keras.models.load_model('LSTM_Attention_Best_model2')

In [27]:
#eval the models
# Evaluate the LSTM model
loss, accuracy = LSTM_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the Bidirectional LSTM model
loss, accuracy = LSTM_Bi_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Between_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention in Between model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Best_model_unclean.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM with Attention Best hyperparameters model - Loss: {loss}, Accuracy: {accuracy}')


5729/5729 [==============================] - 200s 34ms/step - loss: 0.7606 - accuracy: 0.7926
LSTM model - Loss: 0.7605955004692078, Accuracy: 0.792608380317688
5729/5729 [==============================] - 483s 83ms/step - loss: 0.7911 - accuracy: 0.7895
Bidirectional LSTM model - Loss: 0.7911399006843567, Accuracy: 0.7895317077636719
5729/5729 [==============================] - 448s 77ms/step - loss: 0.2097 - accuracy: 0.9109
LSTM with Attention model - Loss: 0.20965580642223358, Accuracy: 0.9109074473381042
5729/5729 [==============================] - 4684s 817ms/step - loss: 0.2087 - accuracy: 0.9122
LSTM with Attention in Between model - Loss: 0.20865486562252045, Accuracy: 0.912222146987915
5729/5729 [==============================] - 688s 116ms/step - loss: 0.8142 - accuracy: 0.7907
LSTM with Attention Best hyperparameters model - Loss: 0.8142003417015076, Accuracy: 0.7906663417816162


| Model                                  | Accuracy | Loss     |
|----------------------------------------|----------|----------|
| LSTM model                             | 0.7926   | 0.7606   |
| Bidirectional LSTM model               | 0.7895   | 0.7911   |
| LSTM with Attention model              | 0.9109   | 0.2097   |
| LSTM with Attention in Between model   | 0.9122   | 0.2087   |
| LSTM with Attention Best hyperparameters model | 0.7907 | 0.8142   |


In [28]:
#eval on clean

loss, accuracy = LSTM_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the Bidirectional LSTM model
loss, accuracy = LSTM_Bi_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Between_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention in Between model - Loss: {loss}, Accuracy: {accuracy}')

# Evaluate the LSTM with Attention model
loss, accuracy = LSTM_Attention_Best_model_unclean.evaluate(X_test_clean, y_test_clean)
print(f'LSTM with Attention Best hyperparameters model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 224s 39ms/step - loss: 2.6503 - accuracy: 0.3347
LSTM model - Loss: 2.6503100395202637, Accuracy: 0.33474621176719666
5729/5729 [==============================] - 414s 72ms/step - loss: 2.5787 - accuracy: 0.3264
Bidirectional LSTM model - Loss: 2.5787415504455566, Accuracy: 0.32639992237091064
5729/5729 [==============================] - 512s 89ms/step - loss: 2.9805 - accuracy: 0.3384
LSTM with Attention model - Loss: 2.980518102645874, Accuracy: 0.33843928575515747
5729/5729 [==============================] - 527s 92ms/step - loss: 2.7619 - accuracy: 0.3380
LSTM with Attention in Between model - Loss: 2.761898994445801, Accuracy: 0.3380083441734314
5729/5729 [==============================] - 568s 99ms/step - loss: 3.0649 - accuracy: 0.3324
LSTM with Attention Best hyperparameters model - Loss: 3.0649240016937256, Accuracy: 0.3324332535266876


| Model                                  | Accuracy | Loss     |
|----------------------------------------|----------|----------|
| LSTM model                             | 0.3347   | 2.6503   |
| Bidirectional LSTM model               | 0.3264   | 2.5787   |
| LSTM with Attention model              | 0.3384   | 2.9805   |
| LSTM with Attention in Between model   | 0.3380   | 2.7619   |
| LSTM with Attention Best hyperparameters model | 0.3324 | 3.0649   |


Lets explore deeper and wider versions

CNN models

In [7]:
#load CNN models
CNN_model_Deep = tf.keras.models.load_model('CNN_modelDeep')
CNN_model_Wide = tf.keras.models.load_model('CNN_modelWide')
CNN_model_Sigmoid_Wide = tf.keras.models.load_model('CNN_model_Sigmoid_Wide')
CNN_model_Sigmoid_Deep = tf.keras.models.load_model('CNN_model_Sigmoid_Deep')
CNN_model_Tanh_Wide = tf.keras.models.load_model('CNN_model_Tanh_Wide')
CNN_model_Tanh_Deep = tf.keras.models.load_model('CNN_model_Tanh_Deep')

In [8]:
#eval the models
loss, accuracy = CNN_model_Deep.evaluate(X_test_clean, y_test_clean)
print(f'Deep CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Wide.evaluate(X_test_clean, y_test_clean)
print(f'Wide CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Sigmoid_Wide.evaluate(X_test_clean, y_test_clean)
print(f'Wide CNN model with Sigmpoid accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Sigmoid_Deep.evaluate(X_test_clean, y_test_clean)
print(f'Deep CNN model with Sigmpoid accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Tanh_Wide.evaluate(X_test_clean, y_test_clean)
print(f'Wide CNN model with Tanh accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Tanh_Deep.evaluate(X_test_clean, y_test_clean)
print(f'Deep CNN model with Tanh accuracy: {accuracy}, loss: {loss}')



5729/5729 [==============================] - 16s 3ms/step - loss: 0.2444 - accuracy: 0.9050
Deep CNN model with Relu accuracy: 0.9049614071846008, loss: 0.24435481429100037
5729/5729 [==============================] - 23s 4ms/step - loss: 0.2679 - accuracy: 0.9043
Wide CNN model with Relu accuracy: 0.904323160648346, loss: 0.2678716778755188
5729/5729 [==============================] - 22s 4ms/step - loss: 0.2693 - accuracy: 0.8994
Wide CNN model with Sigmpoid accuracy: 0.8994244933128357, loss: 0.2692873775959015
5729/5729 [==============================] - 14s 2ms/step - loss: 0.2383 - accuracy: 0.9053
Deep CNN model with Sigmpoid accuracy: 0.9052832722663879, loss: 0.23827773332595825
5729/5729 [==============================] - 20s 3ms/step - loss: 0.2656 - accuracy: 0.9007
Wide CNN model with Tanh accuracy: 0.9006573557853699, loss: 0.2655893862247467
5729/5729 [==============================] - 15s 3ms/step - loss: 0.2428 - accuracy: 0.9018
Deep CNN model with Tanh accuracy: 0.90

| Model                     | Accuracy | Loss     |
|---------------------------|----------|----------|
| Deep CNN with Relu        | 0.9050   | 0.2444   |
| Wide CNN with Relu        | 0.9043   | 0.2679   |
| Wide CNN with Sigmoid     | 0.8994   | 0.2693   |
| Deep CNN with Sigmoid     | 0.9053   | 0.2383   |
| Wide CNN with Tanh        | 0.9007   | 0.2656   |
| Deep CNN with Tanh        | 0.9018   | 0.2428   |


In [9]:
#eval on unclean

loss, accuracy = CNN_model_Deep.evaluate(X_test_unclean, y_test_unclean)
print(f'Deep CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Wide.evaluate(X_test_unclean, y_test_unclean)
print(f'Wide CNN model with Relu accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Sigmoid_Wide.evaluate(X_test_unclean, y_test_unclean)
print(f'Wide CNN model with Sigmpoid accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Sigmoid_Deep.evaluate(X_test_unclean, y_test_unclean)
print(f'Deep CNN model with Sigmpoid accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Tanh_Wide.evaluate(X_test_unclean, y_test_unclean)
print(f'Wide CNN model with Tanh accuracy: {accuracy}, loss: {loss}')

loss, accuracy = CNN_model_Tanh_Deep.evaluate(X_test_unclean, y_test_unclean)
print(f'Deep CNN model with Tanh accuracy: {accuracy}, loss: {loss}')



5729/5729 [==============================] - 14s 2ms/step - loss: 6.7808 - accuracy: 0.3654
Deep CNN model with Relu accuracy: 0.365447461605072, loss: 6.780791282653809
5729/5729 [==============================] - 20s 3ms/step - loss: 7.2220 - accuracy: 0.3639
Wide CNN model with Relu accuracy: 0.36390912532806396, loss: 7.221972942352295
5729/5729 [==============================] - 21s 4ms/step - loss: 4.3606 - accuracy: 0.3608
Wide CNN model with Sigmpoid accuracy: 0.360816091299057, loss: 4.360630989074707
5729/5729 [==============================] - 16s 3ms/step - loss: 3.7525 - accuracy: 0.3694
Deep CNN model with Sigmpoid accuracy: 0.3694242238998413, loss: 3.7524960041046143
5729/5729 [==============================] - 21s 4ms/step - loss: 4.3768 - accuracy: 0.3649
Wide CNN model with Tanh accuracy: 0.3648964762687683, loss: 4.376848220825195
5729/5729 [==============================] - 16s 3ms/step - loss: 3.7347 - accuracy: 0.3669
Deep CNN model with Tanh accuracy: 0.36691486

| Model                     | Accuracy | Loss     |
|---------------------------|----------|----------|
| Deep CNN with Relu        | 0.3654   | 6.7808   |
| Wide CNN with Relu        | 0.3639   | 7.2220   |
| Wide CNN with Sigmoid     | 0.3608   | 4.3606   |
| Deep CNN with Sigmoid     | 0.3694   | 3.7525   |
| Wide CNN with Tanh        | 0.3649   | 4.3768   |
| Deep CNN with Tanh        | 0.3669   | 3.7347   |


LSTM Models

In [11]:
LSTM_Bi_model_Deep = tf.keras.models.load_model('LSTM_Bi_modelDeep')
LSTM_Bi_model_Wide = tf.keras.models.load_model('LSTM_Bi_modelWide')
LSTM_Attention_model_Deep = tf.keras.models.load_model('LSTM_Attention_modelDeep')
LSTM_Attention_model_Wide = tf.keras.models.load_model('LSTM_Attention_modelWide')


In [12]:
#eval the models
loss, accuracy = LSTM_Bi_model_Deep.evaluate(X_test_clean, y_test_clean)
print(f'Deep Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Bi_model_Wide.evaluate(X_test_clean, y_test_clean)
print(f'Wide Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Attention_model_Deep.evaluate(X_test_clean, y_test_clean)
print(f'Deep Attention LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Attention_model_Wide.evaluate(X_test_clean, y_test_clean)
print(f'Wide Attention LSTM model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 135s 23ms/step - loss: 0.2138 - accuracy: 0.9095
Deep Bidirectional LSTM model - Loss: 0.21381835639476776, Accuracy: 0.9094836711883545
5729/5729 [==============================] - 213s 37ms/step - loss: 0.2097 - accuracy: 0.9104
Wide Bidirectional LSTM model - Loss: 0.20974062383174896, Accuracy: 0.910389244556427
5729/5729 [==============================] - 519s 90ms/step - loss: 0.2140 - accuracy: 0.9083
Deep Attention LSTM model - Loss: 0.2139802873134613, Accuracy: 0.9082617163658142
5729/5729 [==============================] - 631s 110ms/step - loss: 0.2133 - accuracy: 0.9086
Wide Attention LSTM model - Loss: 0.21325065195560455, Accuracy: 0.9086217880249023


| Model                        | Accuracy | Loss    |
|------------------------------|----------|---------|
| Deep Bidirectional LSTM      | 0.9095   | 0.2138  |
| Wide Bidirectional LSTM      | 0.9104   | 0.2097  |
| Deep Attention LSTM          | 0.9083   | 0.2140  |
| Wide Attention LSTM          | 0.9086   | 0.2133  |


In [13]:
#eval on unclean
loss, accuracy = LSTM_Bi_model_Deep.evaluate(X_test_unclean, y_test_unclean)
print(f'Deep Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Bi_model_Wide.evaluate(X_test_unclean, y_test_unclean)
print(f'Wide Bidirectional LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Attention_model_Deep.evaluate(X_test_unclean, y_test_unclean)
print(f'Deep Attention LSTM model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = LSTM_Attention_model_Wide.evaluate(X_test_unclean, y_test_unclean)
print(f'Wide Attention LSTM model - Loss: {loss}, Accuracy: {accuracy}')

   1/5729 [..............................] - ETA: 4:05 - loss: 4.5642 - accuracy: 0.2812

5729/5729 [==============================] - 135s 23ms/step - loss: 3.7892 - accuracy: 0.3672
Deep Bidirectional LSTM model - Loss: 3.7891995906829834, Accuracy: 0.3672312796115875
5729/5729 [==============================] - 227s 40ms/step - loss: 3.7924 - accuracy: 0.3654
Wide Bidirectional LSTM model - Loss: 3.79241681098938, Accuracy: 0.36540380120277405
5729/5729 [==============================] - 556s 97ms/step - loss: 3.1003 - accuracy: 0.3576
Deep Attention LSTM model - Loss: 3.100304126739502, Accuracy: 0.3575811982154846
5729/5729 [==============================] - 670s 117ms/step - loss: 3.4582 - accuracy: 0.3661
Wide Attention LSTM model - Loss: 3.458219528198242, Accuracy: 0.3660966157913208


| Model                        | Accuracy | Loss    |
|------------------------------|----------|---------|
| Deep Bidirectional LSTM      | 0.3672   | 3.7892  |
| Wide Bidirectional LSTM      | 0.3654   | 3.7924  |
| Deep Attention LSTM          | 0.3576   | 3.1003  |
| Wide Attention LSTM          | 0.3661   | 3.4582  |


Now taking the best LSTM and the best CNN lets make 2 conbined models and see if there is a difference

In [14]:
LSTM_CNN_model = tf.keras.models.load_model('LSTM_CNN_model')
CNN_LSTM_model = tf.keras.models.load_model('CNN_LSTM_model')

In [15]:
#eval the models
loss, accuracy = LSTM_CNN_model.evaluate(X_test_clean, y_test_clean)
print(f'LSTM CNN model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = CNN_LSTM_model.evaluate(X_test_clean, y_test_clean)
print(f'CNN LSTM model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 239s 41ms/step - loss: 0.2067 - accuracy: 0.9115
LSTM CNN model - Loss: 0.20674003660678864, Accuracy: 0.9114911556243896
5729/5729 [==============================] - 31s 5ms/step - loss: 0.2200 - accuracy: 0.9071
CNN LSTM model - Loss: 0.21999554336071014, Accuracy: 0.9071106910705566


In [16]:
#unlean
loss, accuracy = LSTM_CNN_model.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM CNN model - Loss: {loss}, Accuracy: {accuracy}')

loss, accuracy = CNN_LSTM_model.evaluate(X_test_unclean, y_test_unclean)
print(f'CNN LSTM model - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 235s 41ms/step - loss: 3.9538 - accuracy: 0.3658
LSTM CNN model - Loss: 3.953833818435669, Accuracy: 0.3657638430595398
5729/5729 [==============================] - 29s 5ms/step - loss: 3.5841 - accuracy: 0.3647
CNN LSTM model - Loss: 3.5841047763824463, Accuracy: 0.3647491931915283


Trained the LSTM CNN on unclean data and clean data, will this improve bith ot just one

In [19]:
LSTM_CNN_model_Final = tf.keras.models.load_model('LSTM_CNN_model_unclean')

In [20]:
#eval the model
loss, accuracy = LSTM_CNN_model_Final.evaluate(X_test_unclean, y_test_unclean)
print(f'LSTM CNN model on unclean data - Loss: {loss}, Accuracy: {accuracy}')

#eval on clean
loss, accuracy = LSTM_CNN_model_Final.evaluate(X_test_clean, y_test_clean)
print(f'LSTM CNN model on clean data - Loss: {loss}, Accuracy: {accuracy}')

5729/5729 [==============================] - 256s 44ms/step - loss: 0.2046 - accuracy: 0.9114
LSTM CNN model on unclean data - Loss: 0.20464429259300232, Accuracy: 0.9113875031471252
5729/5729 [==============================] - 255s 44ms/step - loss: 0.2102 - accuracy: 0.9095
LSTM CNN model on clean data - Loss: 0.21021819114685059, Accuracy: 0.9094727635383606
